File: public/packages/dotnet/messaging/rabbitmq/Topology/DlqNaming.cs
// -----------------------------------------------------------------------
// <copyright file="DlqNaming.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Messaging.RabbitMq.Topology;

/// <summary>
/// Conventional names for the dead-letter exchange / queue that pair with
/// a primary queue, plus retry-tier naming used by the optional broker-level
/// retry topology.
/// </summary>
/// <remarks>
/// The convention is uniform across all queues so ops tooling
/// (<c>d2 msg decrypt --queue {q}.dlq</c>) and dashboards can rely on a
/// fixed shape. The lib does not allow overriding these names — they're
/// derived from the queue name, full stop.
/// </remarks>
internal static class DlqNaming
{
    /// <summary>Returns the DLX name for a primary queue.</summary>
    /// <param name="queueName">Primary queue name.</param>
    public static string DlxFor(string queueName) => $"{queueName}.dlx";

    /// <summary>Returns the DLQ name for a primary queue.</summary>
    /// <param name="queueName">Primary queue name.</param>
    public static string DlqFor(string queueName) => $"{queueName}.dlq";

    /// <summary>Returns the retry-tier exchange name (one per tier).</summary>
    /// <param name="queueName">Primary queue name.</param>
    /// <param name="tierIndex">Zero-based tier index.</param>
    public static string RetryTierExchangeFor(string queueName, int tierIndex)
        => $"{queueName}.retry.{tierIndex}";

    /// <summary>Returns the retry-tier queue name (one per tier).</summary>
    /// <param name="queueName">Primary queue name.</param>
    /// <param name="tierIndex">Zero-based tier index.</param>
    public static string RetryTierQueueFor(string queueName, int tierIndex)
        => $"{queueName}.retry.{tierIndex}";

    /// <summary>Returns the return exchange name — retry queues dead-letter
    /// here on TTL expiry; binding routes back to the primary queue.</summary>
    /// <param name="queueName">Primary queue name.</param>
    public static string RetryReturnExchangeFor(string queueName)
        => $"{queueName}.retry.return";
}

File: public/packages/typescript/messaging/rabbitmq/src/topology/dlq-naming.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

/**
 * Conventional names for the dead-letter exchange / queue that pair with a
 * primary queue, plus retry-tier naming used by the optional broker-level
 * retry topology. Byte-identical to the .NET
 * `D2.Shared.Messaging.RabbitMq.Topology.DlqNaming` — the names are derived
 * from the queue name, full stop (no override), so ops tooling and dashboards
 * rely on a fixed shape across both runtimes.
 */
export const DlqNaming = {
  /** Returns the DLX name for a primary queue. */
  dlxFor(queueName: string): string {
    return `${queueName}.dlx`;
  },

  /** Returns the DLQ name for a primary queue. */
  dlqFor(queueName: string): string {
    return `${queueName}.dlq`;
  },

  /** Returns the retry-tier exchange name (one per tier). */
  retryTierExchangeFor(queueName: string, tierIndex: number): string {
    return `${queueName}.retry.${tierIndex}`;
  },

  /** Returns the retry-tier queue name (one per tier). */
  retryTierQueueFor(queueName: string, tierIndex: number): string {
    return `${queueName}.retry.${tierIndex}`;
  },

  /**
   * Returns the return exchange name — retry queues dead-letter here on TTL
   * expiry; the binding routes back to the primary queue.
   */
  retryReturnExchangeFor(queueName: string): string {
    return `${queueName}.retry.return`;
  },
} as const;
